incubator/hnc/pkg/reconcilers/hierarchy_config.go

```diff
@@ -296,7 +296,25 @@ func (r *HierarchyConfigReconciler) syncParent(log logr.Logger, inst *api.Hierar
 // flushObsoleteObjectConditions looks for object conditions from objects that were in one of the
 // namespaces that's been removed as a descendant, and enqueues any affected namespaces.
 func (r *HierarchyConfigReconciler) flushObsoleteObjectConditions(log logr.Logger, ns, oldAns *forest.Namespace) {
-	// Create the set of all relevant names
+	// There should be no obsolete object conditions if the old parent is empty.
+	if oldAns == nil {
+		return
+	}
+
+	// To clear all descendants conditions introduced by ancestors, create a set of all ancestor names.
+	wasAnc := map[string]bool{}
+	for _, anc := range oldAns.AncestryNames(nil) {
+		wasAnc[anc] = true
+	}
+
+	// Remove conditions referring to its ancestors from all its descendants.
+	for _, desc := range ns.DescendantNames() {
```

Comment:
You may as well build the `wasDesc` set in this loop, instead of having a separate loop

Reply:
sg, combined!